### ISSUE TITLE
db_bench usability of the pinning policy parameter 

### ISSUE THREAD

**Issue by erez-speedb** [speedb-io/speedb#687]
The current parameter is undocumented and complicated:
-pinning_policy=""id=speedb_scoped_pinning_policy;capacity=$(($cache_size * 80 /100))"

I suggest something simple like:
-pinning_policy=speedb_scoped_pinning_policy;
-pinnig_capacity= % from block cache

This should also imply -cache_index_and_filter_blocks

**Comment by erez-speedb** [comment]
@udi-speedb  @hilikspdb  based on what were those numbers chosen?

**Comment by udi-speedb** [comment]
I am going to do the following:

1. Cache size is set independently by the user (or implicitly). db_bench should do the calculations on its own rather than expect the user to provide a number.
2. I will take into account the WBM if enabled and configured to cost to cache. So, the pinning policy will be configured to use capacity = 80% of (cache-size - WBM (if applicable) )
3. If the user specifies a pinning policy and configured cache_index_and_filter_blocks=false, db_bench will report the mis-configuration and exit.
4. The default pinning policy is to use a policy that should result in the same behavior as the one in RocksDB or that was in effect before static pinning feature was merged (=> pinning based on the multiple existing pinning options). The alternative is to use our own Scoped pinning policy that pins based on the thresholds, ignoring the pinning options. Although the way it is implemented in db_bench is generic, and support any pinning policy, it is inconvenient (that's how this issue came to be). So, I will have the following db_bench flags to control pinning:
- scoped_pinning_policy: Boolean. False by default => Default pinning policy. When true, scoped pinning policy.

- scoped_pinning_last_level_with_data_percent: Sets the bottom threshold (applicable only when scoped_pinning_policy=true). Default: Same as the defaults set in the code (currently 10% but will be synchronized with whatever is set in the code (soon will be 70%)

- scoped_pinning_mid_percent: Sets the middle (non-level-0) threshold (applicable only when scoped_pinning_policy=true).  Default: Same as the defaults set in the code (currently 10% but will be synchronized with whatever is set in the code (soon will be 90%)

scoped_pinning_mid_percent must be >= scoped_pinning_last_level_with_data_percent

**Comment by udi-speedb** [comment]
> @udi-speedb @hilikspdb based on what were those numbers chosen?

Which numbers? Should this question be asked in #689 ?

**Comment by mrambacher** [comment]
I agree with the sentiment of this request and agree that percentages should be possible.  This is something however that requires more communication in the infrastructure -- you need to know what the Cache size is for the pinning -- and would need to take place in something like PrepareOptions.  

Note that #376 is a more generic attempt to reconcile this problem where options of sub options (like block size to table factory) could be configured in a standard way for all of these options.  For this specific case, the idea would be to have two options like "--pinning_policy=speedb --mid_percent=70", which would be cleaner and simpler than the complex form currently required.

**Comment by udi-speedb** [comment]
The feature is currently in the release and we need a simple and convenient way to use it. 
I understand that the heuristics is 80% of the available cache size (accounting for the WBM is applicable). So, I really don't understand the problem with my suggestion. The code is simple and it would provide @erez-speedb as well as others with simple and convenient means to configure it. 
Having superior generic mechanisms is great, but as long as they are not available, I'd rather have what I have suggested.

**Comment by mrambacher** [comment]
> The feature is currently in the release and we need a simple and convenient way to use it. I understand that the heuristics is 80% of the available cache size (accounting for the WBM is applicable). So, I really don't understand the problem with my suggestion. The code is simple and it would provide @erez-speedb as well as others with simple and convenient means to configure it. Having superior generic mechanisms is great, but as long as they are not available, I'd rather have what I have suggested.

I am not sure how your suggestion works for all tools and users and not just db_bench.  My suggestion was how to make db_bench (and the other tools) easier to configure and was not specific to this issue.

If there is a suggestion to create a use case that sets these parameters generically for all Speedb users, that would be great but would be bigger than the scope of (at least the title) of this issue.

**Comment by udi-speedb** [comment]
My suggestion is clearly limited to db_bench. The title of this issue indicates it.

I am trying to provide the users of db_bench simple means to use the feature and easily configure it, as it is currently implemented. This is essential so we are able to test the feature **now**.

**Comment by udi-speedb** [comment]
@erez-speedb - Are you satisfied with the way db_bench supports pinning configuration?

**Comment by erez-speedb** [comment]
As design, the flag requires caching index and filters
`ERROR: --cache_index_and_filter_blocks must be set for --scoped_pinning_policy to have any affect.`
@udi-speedb  can you set that the scoped_pinning_policy will set the cache index automatically?

**Comment by erez-speedb** [comment]
Test was run with parameters 
` -scoped_pinning_mid_percent=70 -scoped_pinning_last_level_with_data_percent=10 -scoped_pinning_policy --cache_index_and_filter_blocks -cache_size=33554432`

The log state as expected 
`pinning_policy_name: speedb_scoped_pinning_policy
  pinning_policy_options:
    pinning_policy.capacity: 26843545
    pinning_policy.last_level_with_data_percent: 10
    pinning_policy.mid_percent: 70`
@udi-speedb do we have a parameter to set the total scope capacity? Or is it always 80

**Comment by udi-speedb** [comment]
> Test was run with parameters ` -scoped_pinning_mid_percent=70 -scoped_pinning_last_level_with_data_percent=10 -scoped_pinning_policy --cache_index_and_filter_blocks -cache_size=33554432`
> 
> The log state as expected `pinning_policy_name: speedb_scoped_pinning_policy pinning_policy_options: pinning_policy.capacity: 26843545 pinning_policy.last_level_with_data_percent: 10 pinning_policy.mid_percent: 70` @udi-speedb do we have a parameter to set the total scope capacity? Or is it always 80

There is **no** separate flag to control the capacity. It is set automatically as follows:
If using a wbm with cost-to-cache:
pinning-capacity = 0.8 * (cache size - wbm buffer size)

Otherwise:
pinning-capacity = 0.8 * cache size

**I will add another flag** (_pinning_capacity_):
- Default will be -1 => Calculate automatically as I have just described
- If the value is >= 0, the pinning capacity will be set to the value of _pinning_capacity_.
